refactor(semantic): always use `SymbolFlags::Function` for function id

Dunqing opened the pull request:
close #7402 
related: #7470

The purpose of this PR is to keep the SymbolFlags of the function consistent.

Always use `SymbolFlags::Function` for the function id symbol, to avoid adding extra logic to find accurate `SymbolFlags` for the function in `transformer`. This change adds a fallback redeclaration check for case `async function foo() {}; var foo;` in `SemanticBuilder::check_redeclaraion`. 

No performance difference here because redeclaration is always rare.

Dunqing commented:
* **#7479** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7479?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7479?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Nov 28, 5:54 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
Can we possibly discuss this before merging? Personally, I'm not convinced this is a good idea.

Boshen commented:
https://github.com/oxc-project/oxc/blob/58a125ff8812e55dba20982b9755b522159431bf/crates/oxc_transformer/src/es2017/async_to_generator.rs#L368-L373

Dunqing committed:
refactor(semantic): always use `SymbolFlags::Function` for function id (#7479)

close #7402
related: #7470

The purpose of this PR is to keep the SymbolFlags of the function consistent.

Always use `SymbolFlags::Function` for the function id symbol, to avoid adding extra logic to find accurate `SymbolFlags` for the function in `transformer`. This change adds a fallback redeclaration check for case `async function foo() {}; var foo;` in `SemanticBuilder::check_redeclaraion`.

No performance difference here because redeclaration is always rare.
